libs/grout/README.md

```diff
@@ -0,0 +1,178 @@
+# Grout
+
+A library to create type-safe RPC glue code.
+
+Grout lets you define a server API as a plain TS object with async methods.
+Based on this API definition, Grout will create Express route handlers on the
+server side and a type-safe client on the client side, abstracting away the
+details of the underlying HTTP requests.
+
+## Installation
+
+Grout must be installed as a dependency in both your client and server projects.
+
+## Usage
+
+### Server
+
+```ts
+import * as grout from '@votingworks/grout';
+
+interface Person {
+  name: string;
+  age: number;
+}
+
+// Create an API definition by building an object with async methods.
+const api = grout.createApi({
+  // Each method should return a Promise. The data returned should be simple TS
+  // values - not fancy class instances. Values that can be easily serialized to
+  // JSON. More details on this below.
+  async getAllPeople(): Promise<Person[]> {
+    return store.getAllPeople(); // Assume this accesses the database, etc.
+  },
+
+  // Methods can take input, but it must be packaged into a single object.
+  // Think of it like using named parameters, or React component props.
+  async getPersonByName(input: { name: string }): Promise<Person | undefined> {
+    return store.getPersonByName(input.name);
+  },
+
+  // For known errors (e.g. invalid input), use a Result type so that the client
+  // forced to handle them explicitly.
+  async updatePersonAge(input: {
+    name: string;
+    age: number;
+  }): Promise<Result<void, Error>> {
+    if (age <= 0) {
+      return err(new Error('Age must be at least 1.'));
+    }
+    await store.updatePersonAge(input.name, input.age);
+    return ok();
+  },
+});
+
+const app = express();
+
+// When registering the API with an Express app, you must use a separate
+// Router. That way you can isolate the Grout API routes from other routes in
+// your app.
+const router = express.Router();
+
+// You must have the router use the `express.text()` middleware with type
+// `application/json` exactly as shown below.
+router.use(express.text({ type: 'application/json' }));
+
+// Then, simply register the API routes.
+grout.registerRoutes(router, api);
+
+// Finally, mount the router on your app. You can use any path prefix you want -
+// here we use `/api`. It's important that you don't use any other body-parsing
+// middleware upstream of the router - e.g. don't call `app.use(express.json())`
+// before this.
+app.use('/api', router);
```

Comment:
Yeah, I did start with that. But then grout has to depend on express. I was a bit wary of ending up with mismatched express versions between grout and its consumers. I considered peer dependencies but was intimidated because they seem finicky. So that's how I landed here. But I'm very willing to reconsider and would love to hear your thoughts about the tradeoffs.

Reply:
Maybe a `peerDependencies` entry would fix that.

Reply:
I just remembered that the other reason I did it this way was to avoid frontends having to have express as a dependency. Another solution to that would be breaking grout into two separate packages - one for frontend and one for backend. Not sure if there's an easy way to do that other than creating two libs (maybe a custom build process?)

On the other hand maybe it doesn't matter and express can just be a peer dep for frontends as well, just seems a little weird.

Reply:
I think the inversion is fine, I just wonder if we can make it _one_ thing you have to do rather than this series of steps.

Reply:
Maybe just passing in the `express` object would solve that.

Reply:
This approach seems to work well: 461034c82. Thanks for pushing on this point!